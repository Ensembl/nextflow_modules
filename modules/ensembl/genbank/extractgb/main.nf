// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

process GENBANK_EXTRACTGB {
    tag "${meta.accession}:${gb_file}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "ensemblorg/ensembl-genomio:GenomioDockerRebuild_v1.5.0a"

    input:
        tuple val(meta), path(gb_file)

    output:
        tuple val(meta), path("genome.json"), emit: genome
        tuple val(meta), path("seq_region.json"), emit: seq_regions
        tuple val(meta), path("dna.fasta"), emit: dna_fasta
        tuple val(meta), path("*.gff"), emit: gene_gff, optional: true
        tuple val(meta), path("pep.fasta"), emit: pep_fasta, optional: true
        path "versions.yml", emit: versions

    when:
        task.ext.when == null || task.ext.when


    shell:
        def args = task.ext.args ?: ''
        def prefix = task.ext.prefix ?: "${gb_file}"

        '''
        genbank_extract_data \
            --prefix !{meta.prefix} \
            --prod_name !{meta.production_name} \
            --gb_file !{gb_file} \
            --debug

        schemas_json_validate --json_file "genome.json" --json_schema "genome"
        schemas_json_validate --json_file "seq_region.json" --json_schema "seq_region"
        '''

    stub:
        def args = task.ext.args ?: ''
        def prefix = task.ext.prefix ?: "${gb_file}"
        out_genome_json = "genome.json"
        out_seq_json = "seq_region.json"
        out_dna_fa = "dna.fasta"
        out_gene_gff = "valida.gff"
        out_pep_fa = "pep.fasta"

        """
        echo "{"accession": "NC_00001"}" > ${out_genome_json}
        echo "{"seq_region": "Chrm1"}" > ${out_seq_json}
        echo -e -n ">Seq1\nATGC" > ${out_dna_fa}
        cp ${workflow.projectDir}/tests/modules/ensembl/download/genbank/test.gff ${out_gene_gff}
        echo ">Peptide_1a\nMPLEGM" > ${out_pep_fa}

        echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
        python -c "import ensembl.io.genomio; print(ensembl.io.genomio.__version__)" >> versions.yml
        """
}
