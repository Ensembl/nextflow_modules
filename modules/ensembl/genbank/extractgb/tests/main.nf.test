// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// nf-core modules test genbank/extractgb
nextflow_process {

    name "Test Process GENBANK_EXTRACTGB"
    script "../main.nf"
    process "GENBANK_EXTRACTGB"

    tag "modules"
    tag "modules_nfcore"
    tag "genbank"
    tag "genbank/extractgb"

test("Genbank extract gb [stub]") {
        when {

            options "-stub"

            process {
                """
                input[0] = [[ accession:'test_genbank'] , file("test_output.gb")]
                """
            }
        }

        then {
            
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.size() >= 3  },
                { assert process.out.genome.size() == 1 },
                { assert process.out.seq_regions.size() == 1 },
                { assert process.out.dna_fasta.size() == 1 },
                { assert process.out.gene_gff.size() <= 1 },
                { assert process.out.pep_fasta.size() <= 1 },
                { assert process.out != null }
            )
        }
    }
}