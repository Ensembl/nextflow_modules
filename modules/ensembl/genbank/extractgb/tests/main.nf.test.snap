{
    "Genbank extract gb [stub]": {
        "content": [
            {
                "0": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "genome.json:md5,d453eccd0182cf9fd072daf4170ca743"
                    ]
                ],
                "1": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "seq_region.json:md5,238b822b713730b11f790ef8fa6372ed"
                    ]
                ],
                "2": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "dna.fasta:md5,7bf591c1a378a9bc6779aa8efd4b94d0"
                    ]
                ],
                "3": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "valida.gff:md5,9fecaee47b109cd504f7a5a217132974"
                    ]
                ],
                "4": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "pep.fasta:md5,71f0789bc9da7799197739226d951d23"
                    ]
                ],
                "5": [
                    "versions.yml:md5,4b0af6aac956a5b0c26acd921a939796"
                ],
                "dna_fasta": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "dna.fasta:md5,7bf591c1a378a9bc6779aa8efd4b94d0"
                    ]
                ],
                "gene_gff": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "valida.gff:md5,9fecaee47b109cd504f7a5a217132974"
                    ]
                ],
                "genome": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "genome.json:md5,d453eccd0182cf9fd072daf4170ca743"
                    ]
                ],
                "pep_fasta": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "pep.fasta:md5,71f0789bc9da7799197739226d951d23"
                    ]
                ],
                "seq_regions": [
                    [
                        {
                            "accession": "test_genbank"
                        },
                        "seq_region.json:md5,238b822b713730b11f790ef8fa6372ed"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4b0af6aac956a5b0c26acd921a939796"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-19T16:14:50.053484519"
    }
}